### PR TITLE
Update and give access to invisible automated testing page doc

### PR DESCRIPTION
**Pull request by plocket**
Addresses #135 for the deep dive if we get there.

This avoids putting the page in the sidebar, but it will be accessible from https://suffolklitlab.org/docassemble-AssemblyLine-documentation/docs/automated_integrated_testing so that we can look at it together during during the deep dive. It updates it and condenses it a bit.

### CODE REVIEW

File: docs/automated_testing.md [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+Secrets
+
+PLAYGROUND_PASSWORD
+PLAYGROUND_ID
+PLAYGROUND_EMAIL
+-->
+
+## Summary
+
+1. You write and edit `.feature` test files in your Sources folder.
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:29]:
```suggestion
1. You write and edit `.feature` test files in your Sources folder that contain a series of steps that the test should take when filling out the interview.
```

File: docs/automated_testing.md [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+1. There's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet.
+1. Tests are run when you commit.
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:31]:
```suggestion
1. Tests are run when you commit (there's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet).
```

File: docs/automated_testing.md [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    Given I start the interview at "plaintiffs_motion_to_modify_209a"
+```
+
+More complex tests might wait till your code is pretty much how you want it. Every time you change your variable names, you may have to update the tests.
+
+## Story tables syntax
+
+Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.
+
+Tip: it can be easy to lose track of what variables you've included, so it can be useful keep the rows in alphabetical order.
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:84]:
```suggestion
:::tip 
It can be easy to lose track of what variables you've included, so it can be useful keep the rows in alphabetical order.
:::
```

File: docs/automated_testing.md [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+- `users[i].address_fields()` generates all the address fields with the right index number (NOT 'i'). E.g. `users[0].address.address`, `users[1].address.address` -->
+
+### value
+
+In the **value** column, write what you want the field to be set to. For checkboxes, `true` means 'checked' and `false` means 'unchecked'.
+
+
+### trigger
+
+In the **trigger** column, write the name of the intrinsic name of the variable that will trigger the page for this field. This column is optional in most cases. In cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object).
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:122]:
```suggestion
In the **trigger** column, write the name of the intrinsic name of the variable that will trigger the page for this field. This column is necessary in cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object), and optional everywhere else.
```

File: docs/automated_testing.md [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+The Step for the story table, which is better described in sections above.
+```
+    I get to the question id "some yml block id" with this data:
+```
+
+### Other actions
+
+```
+    When I tap to continue
+```
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:283]: This one's duplicated?

File: docs/automated_testing.md [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+
+## Test instructions details
+
+### Add a new test
+
+Go to your Playground > the dropdown Folders menu > Sources.
+
+Add a new file that ends in the extension `.feature`. Ex: `has_children.feature`
+
+The next time you commit, your test will be run. You can add and edit multiple test files.
```
Comment (BryceStevenWilley) [inline docs/automated_testing.md:401]: Is setting up the github action part of the first section, "Setup", that's commented out?

File: docs/automated_testing.md [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+
+- Have admin permissions on the repo
+- Use the setup interview (https://apps-dev.suffolklitlab.org/start/test-setup/)
+- Add this element to the page somewhere: <div data-variable="${ encode_name(str( user_info().variable )) }" id="trigger" aria-hidden="true" style="display: none;"></div>
+
+Secrets
+
+PLAYGROUND_PASSWORD
+PLAYGROUND_ID
+PLAYGROUND_EMAIL
```
[the remark this answers is not in the document]
Comment (plocket) [reply docs/automated_testing.md:24]: This info is actually not 'your' info, it's the info of the testing account. This info is only needed for local development. Online, the setup interview walks people through putting in the right info.

[the remark this answers is not in the document]
Comment (plocket) [reply docs/automated_testing.md:24]:
```suggestion
If you run the tests locally, add these environment variables to your `.env` file:

 -` PLAYGROUND_EMAIL`: the docassemble server email for the account that is running the tests.
- `PLAYGROUND_PASSWORD`: the docassemble server password for the account that is running the tests.
- `PLAYGROUND_ID`: the id that shows up in interviews when the testing account runs the tests from the server playground.
```

File: docs/automated_testing.md (hunk 1 above)
Comment (plocket) [reply docs/automated_testing.md:29]: I really want this to be a very bare bones summary. A very quick reference. How about this?

```suggestion
1. You write and edit `.feature` test files in your Sources folder. Instructions for writing tests are below.
```

File: docs/automated_testing.md (hunk 2 above)
Comment (plocket) [reply docs/automated_testing.md:31]: I'd like each item to be as short and well-contained as possible. I think the two concepts are distinct and unrelated. I only mention the timeout because I think it's pretty crucial. How about just switching the order?

```suggestion
1. Tests are run when you commit.
1. There's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet.
```

File: docs/automated_testing.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (BryceStevenWilley) [reply docs/automated_testing.md:24]: Thanks for the correction, and exactly the reason this short clarifier would help a lot, including me! :sweat_smile:

File: docs/automated_testing.md (hunk 1 above)
Comment (BryceStevenWilley) [reply docs/automated_testing.md:29]: I guess, but the issue I had with this line is that if you don't already know what a `.feature` file is, it tells you very little. Just knowing that it's some steps for the tester is good.

But this is minor for just the deep-dive version of the code. Feel free to just ignore for now.

File: docs/automated_testing.md [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+
+  Scenario: The Plaintiff's Motion to Modify loads
+    Given I start the interview at "plaintiffs_motion_to_modify_209a"
+```
+
+More complex tests might wait till your code is pretty much how you want it. Every time you change your variable names, you may have to update the tests.
+
+## Story tables syntax
+
+Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.
```
[the remark this answers is not in the document]
Comment (plocket) [reply docs/automated_testing.md:82]: Very good observation of missing information. Let me know if the below offers more clarity.

```suggestion
Very basically, a **story table** contains a list of variables that correspond to fields and the values you want for those fields. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.
```

File: docs/automated_testing.md (hunk 2 above)
Comment (BryceStevenWilley) [reply docs/automated_testing.md:31]: Taking a step back I guess, what are these items? They read more like steps that the test author takes, but the "There's a maximum of 2 minutes per individual Step..." item isn't something you can do. The second item probably doesn't need to be mentioned here, and can be moved later, even if it's crucial (i.e., is it information that someone, encountering the testing system for the first time, would even remember by the time they got to writing tests?)

File: docs/automated_testing.md (hunk 4 above)
Comment (plocket) [reply docs/automated_testing.md:124]: I like it and I'll add one improvement where I see what I think is a typo of mine and another clarification:

```suggestion
In the **trigger** column, write the intrinsic name of the variable that will trigger the page where this field appears. This column is necessary in cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object), and optional everywhere else.
```

File: docs/automated_testing.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (BryceStevenWilley) [reply docs/automated_testing.md:82]: Adding fields into the mix makes it more difficult to understand I think, mostly because you now have to distinguish between web fields (which these are) and PDF fields, and juggle 3 concepts around (fields, variables, and values) instead of just two (variables and values). The original might actually not be that bad now that I'm looking at it again. :shrug:

File: docs/automated_testing.md (hunk 5 above)
Comment (plocket) [reply docs/automated_testing.md:285]: Yes. I wasn't sure whether to duplicate it or not - duplication is generally bad, but it's one I've used a lot under some circumstances, so I want it to be easy to find and it does seem to be to belong in both categories. What do you think?

Comment (BryceStevenWilley) [reply docs/automated_testing.md:285]: That makes sense. IMO the design of this section on Steps might need some more work, but you're right, it's probably fine for now.

File: docs/automated_testing.md (hunk 6 above)
Comment (plocket) [reply docs/automated_testing.md:403]: Yes. In fact, the developer doesn't have to actively set up any actions. They just go through that setup interview. It should be self-explanatory when they run it, but I haven't user tested it yet. That said, the current PR is an MVP for making something to help conversations between the folks in the AL project about writing tests. I can do setup for our repos for now.

Let me know if we should discuss more. If this is a satisfactory answer, feel free to resolve this.

File: docs/automated_testing.md (hunk 1 above)
Comment (plocket) [reply docs/automated_testing.md:29]: I think this is something that the documentation guidelines site might help with - figuring out where different pieces might go.

File: docs/automated_testing.md (hunk 2 above)
Comment (plocket) [reply docs/automated_testing.md:31]: Ah, I see. Can we leave it this way for the MVP and revisit the best place to put this info when re-organizing a la divio? If so, resolve for now and we can make an issue for it to avoid forgetting.

File: docs/automated_testing.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (plocket) [reply docs/automated_testing.md:82]: How about this?

```suggestion
Very basically, a **story table** contains a list of variables and values. When the test is going through the interview and arrives on a page that has a field that defines the variable, the field will be set with the value. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions.
```

File: docs/automated_testing.md (hunk 2 above)
Comment (plocket) [reply docs/automated_testing.md:31]: #138

File: docs/automated_testing.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (BryceStevenWilley) [reply docs/automated_testing.md:82]: That's a great improvement! :+1:

File: docs/automated_testing.md (hunk 6 above)
Comment (BryceStevenWilley) [reply docs/automated_testing.md:403]: That's fine, skimmed right over that sentence, and it explains the part I thought was missing. Thanks!

**Comment by plocket** [comment]
I think I've got them all, but if I've missed anything, let me know and I'll make changes. I have to update one item anyway.

**Review by BryceStevenWilley — APPROVED** [review]
Explicit approval, which I should have given on the first review.

I usually approve when I don't mind if the suggestions are taken or not, which was def the case with this PR.

**Comment by nonprofittechy** [comment]
FYI @plocket @BryceStevenWilley I'm going to find a place to put this on the sidebar so it's more discoverable
